sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityFilter.java

```diff
@@ -0,0 +1,139 @@
+    private final ThreadContext threadContext;
+    private final ClusterService cs;
+    private final Client client;
+
+    public SecurityFilter(final Client client, final Settings settings, ThreadPool threadPool, ClusterService cs) {
+        this.client = client;
+        this.threadContext = threadPool.getThreadContext();
+        this.cs = cs;
+    }
+
+    @Override
+    public int order() {
+        return Integer.MIN_VALUE;
+    }
+
+    @Override
+    public <Request extends ActionRequest, Response extends ActionResponse> void apply(
+        Task task,
+        final String action,
+        Request request,
+        ActionListener<Response> listener,
+        ActionFilterChain<Request, Response> chain
+    ) {
+        try (StoredContext ctx = threadContext.newStoredContext(true)) {
+            org.apache.logging.log4j.ThreadContext.clearAll();
+            apply0(task, action, request, listener, chain);
+        }
+    }
+
+    private <Request extends ActionRequest, Response extends ActionResponse> void apply0(
+        Task task,
+        final String action,
+        Request request,
+        ActionListener<Response> listener,
+        ActionFilterChain<Request, Response> chain
+    ) {
+        try {
+            // TODO Get jwt here and verify
+            // TODO Move this logic to right after successful login
+            if (threadContext.getHeader(ThreadContextConstants.OPENSEARCH_AUTHENTICATION_TOKEN_HEADER) != null) {
+                String encodedJwt = threadContext.getHeader(ThreadContextConstants.OPENSEARCH_AUTHENTICATION_TOKEN_HEADER);
+                String prefix = "(nodeName="
+                    + cs.localNode().getId()
+                    + ", requestId="
+                    + request.getParentTask().getId()
+                    + ", action="
+                    + action
+                    + " apply0)";
+                log.info(prefix + " Access token provided " + encodedJwt);
+            } else {
+                // TODO Figure out where internal actions are invoked and create token on invocation
+                // No token provided, may be an internal request
+                // Token in ThreadContext is created on REST layer and passed to Transport Layer.
+                String prefix = "(nodeName="
+                    + cs.localNode().getName()
+                    + ", requestId="
+                    + request.getParentTask().getId()
+                    + ", action="
+                    + action
+                    + " apply0)";
+                log.info(prefix + "No authorization provided in the request, internal request");
+                // String err = "Access token not provided";
+                // listener.onFailure(new OpenSearchSecurityException(err, RestStatus.FORBIDDEN));
+            }
+
+            final PrivilegesEvaluatorResponse pres = new PrivilegesEvaluatorResponse(); // eval.evaluate(user, action, request, task,
```

Comment:
What do you think about using `Identity.getSubject().checkPermission(...)`?

Reply:
Yes! When we get more into authz I would love to experiment with the subject APIs.